options/options.c

```diff
@@ -185,53 +198,50 @@ static const m_option_t mp_vo_opt_list[] = {
     {"x11-present", OPT_CHOICE(x11_present,
-        {"no", 0}, {"auto", 1}, {"yes", 2})},
-    {"x11-wid-title", OPT_BOOL(x11_wid_title)},
-#endif
-#if HAVE_WAYLAND
+        {"no", 0}, {"auto", 1}, {"yes", 2}), .not_available = !HAVE_X11},
+    {"x11-wid-title", OPT_BOOL(x11_wid_title), .not_available = !HAVE_X11},
+
+    // Wayland options
     {"wayland-configure-bounds", OPT_CHOICE(wl_configure_bounds,
-        {"auto", -1}, {"no", 0}, {"yes", 1})},
+        {"auto", -1}, {"no", 0}, {"yes", 1}), .not_available = !HAVE_WAYLAND},
     {"wayland-content-type", OPT_CHOICE(wl_content_type, {"auto", -1}, {"none", 0},
-        {"photo", 1}, {"video", 2}, {"game", 3})},
-    {"wayland-disable-vsync", OPT_BOOL(wl_disable_vsync)},
+        {"photo", 1}, {"video", 2}, {"game", 3}), .not_available = !HAVE_WAYLAND},
+    {"wayland-disable-vsync", OPT_BOOL(wl_disable_vsync),
+        .not_available = !HAVE_WAYLAND},
     {"wayland-edge-pixels-pointer", OPT_INT(wl_edge_pixels_pointer),
-        M_RANGE(0, INT_MAX)},
+        M_RANGE(0, INT_MAX), .not_available = !HAVE_WAYLAND},
     {"wayland-edge-pixels-touch", OPT_INT(wl_edge_pixels_touch),
-        M_RANGE(0, INT_MAX)},
-    {"wayland-present", OPT_BOOL(wl_present)},
-#endif
-#if HAVE_WIN32_DESKTOP
-// For old MinGW-w64 compatibility
-#define DWMWCP_DEFAULT 0
-#define DWMWCP_DONOTROUND 1
-#define DWMWCP_ROUND 2
-#define DWMWCP_ROUNDSMALL 3
-
-#define DWMSBT_AUTO 0
-#define DWMSBT_NONE 1
-#define DWMSBT_MAINWINDOW 2
-#define DWMSBT_TRANSIENTWINDOW 3
-#define DWMSBT_TABBEDWINDOW 4
+        M_RANGE(0, INT_MAX), .not_available = !HAVE_WAYLAND},
+    {"wayland-present", OPT_BOOL(wl_present), .not_available = !HAVE_WAYLAND},
 
+    // Windows Desktop options
     {"backdrop-type", OPT_CHOICE(backdrop_type, {"auto", DWMSBT_AUTO}, {"none", DWMSBT_NONE},
-        {"mica", DWMSBT_MAINWINDOW}, {"acrylic", DWMSBT_TRANSIENTWINDOW}, {"mica-alt", DWMSBT_TABBEDWINDOW})},
-    {"window-affinity", OPT_CHOICE(window_affinity, {"default", WDA_NONE},
-        {"excludefromcapture", WDA_EXCLUDEFROMCAPTURE}, {"monitor", WDA_MONITOR})},
-    {"vo-mmcss-profile", OPT_STRING(mmcss_profile)},
+        {"mica", DWMSBT_MAINWINDOW}, {"acrylic", DWMSBT_TRANSIENTWINDOW}, {"mica-alt", DWMSBT_TABBEDWINDOW}),
+        .not_available = !HAVE_WIN32_DESKTOP},
+    {"window-affinity", OPT_CHOICE(window_affinity, {"default", 0},
+        {"excludefromcapture", 0}, {"monitor", WDA_MONITOR}),
+        .not_available = !HAVE_WIN32_DESKTOP},
+    {"vo-mmcss-profile", OPT_STRING(mmcss_profile), .not_available = !HAVE_WIN32_DESKTOP},
     {"window-corners", OPT_CHOICE(window_corners,
         {"default", DWMWCP_DEFAULT},
         {"donotround", DWMWCP_DONOTROUND},
         {"round", DWMWCP_ROUND},
-        {"roundsmall", DWMWCP_ROUNDSMALL})},
-#endif
-#if HAVE_EGL_ANDROID
-    {"android-surface-size", OPT_SIZE_BOX(android_surface_size)},
-#endif
+        {"roundsmall", DWMWCP_ROUNDSMALL}),
+        .not_available = !HAVE_WIN32_DESKTOP},
```

Comment:
Maybe instead of defining all win32 dummy values in meson, we could try to not use them?

``` c
    {"window-corners",
#if HAVE_WIN32_DESKTOP
       OPT_CHOICE(window_corners,
        {"default", DWMWCP_DEFAULT},
        {"donotround", DWMWCP_DONOTROUND},
        {"round", DWMWCP_ROUND},
        {"roundsmall", DWMWCP_ROUNDSMALL})}
#else
       OPT_NOTAVAILABLE}
#endif
```

formatting can be fixed to look better, but point is to avoid using those values. I feel like those dummy definitions are polluting too much

Reply:
We do need to be sure to keep the same valid choices to avoid it erroring out. I could do something like this.

```
#if HAVE_WIN32_DESKTOP
       OPT_CHOICE(window_corners,
        {"default", DWMWCP_DEFAULT},
        {"donotround", DWMWCP_DONOTROUND},
        {"round", DWMWCP_ROUND},
        {"roundsmall", DWMWCP_ROUNDSMALL})}
#else
        OPT_CHOICE_ZERO(window_corners, "default",  "donotround", "round", "roundsmall")
#endif
```

Only bummer with this approach is that the option names have to be typed twice but maybe not such a big deal.

Reply:
> We do need to be sure to keep the same valid choices to avoid it erroring out.

Can't we just skip checking choices for unavailable options?

Reply:
I don't see how we skip setting choices short of skipping the option altogether?

Reply:
I though the point was to show warning when using not compiled in option. So why it maters what args are used, if it is not available anyway?

Reply:
Well the choices could be set to nothing but then that makes them hard errors on the cli again which is inconsistent with the other ones.